### ISSUE TITLE
getAll() method and paging does not seem to be working

### ISSUE THREAD

**Issue by dlegatt** [spinen/connectwise-php-client#40]
https://github.com/spinen/connectwise-php-client/blob/be0b39174704cdd894d3a87b319bbf68934266e3/src/Api/Client.php#L157
When this line is run on the initial request, `$this->page` is actually set to `true` instead of an int. Technically, this would be OK, but the problem arises when my token immediately needs refreshing and then the same client instance is called with a `post` method, at which point `$this->page` is set to `0`. The result is `getAll` only ever returns 1 page of data.

Paging and passing query options has numerous issues that I won't be able to go into right now, but I was able to make it work by changing the ternary expression to this:

```php
if ($verb === 'getAll') {
    $this->page = 1;
    $verb = 'get';
}
```

If `$verb` is anything but `getAll`, it remains null, so the other comparisons still work, and the worst that happens is an unnecessary `page=1` query parameter is passed.

If I have some time next month, I can put together an issue illustrating what I saw with the query/options passing as well as a pull request with a possible solution.

**Comment by jimmypuckett** [comment]
@dlegatt I am sorry to hear that you are having issues.  Here is a small example of interacting with the class in a project that we use the code...

```php
$ docker-compose exec web php artisan tinker
Psy Shell v0.9.9 (PHP 7.2.14 — cli) by Justin Hileman
>>> $cw = app(Spinen\ConnectWise\Api\Client::class)
=> Spinen\ConnectWise\Api\Client {#3558}
>>> count($cw->getAll('/system/members'))
=> 54
>>> count($cw->get('/system/members'))
=> 25
>>> count($cw->get('/system/members?page=2'))
=> 25
>>> count($cw->get('/system/members?page=3'))
=> 4
>>>
```

Can you supply some example of where it is not working, so we can dig into it?

Also, what version of ConnectWise are you connecting & is it hosted?

```php
$ docker-compose exec web php artisan tinker
Psy Shell v0.9.9 (PHP 7.2.14 — cli) by Justin Hileman
>>> $cw = app(Spinen\ConnectWise\Api\Client::class)
=> Spinen\ConnectWise\Api\Client {#3558}
>>> $cw->get('system/info')->version
=> "v2018.6.59996"
>>> $cw->get('system/info')->isCloud
=> false
>>>
```

We have seen some issues in a past versions (don't remember which one) of ConnectWise where sending `page=1` for request that only had 1 page give an error, so we are specifically using `null` for `no page`.  We stopped sending the `page=1` for every request.

Anyhow, I hope that you are able to find some time next month to layout some of the issues, as it would be nice to have someone else improve the client.

Good luck & I look forward in hearing back from you.

**Comment by dlegatt** [comment]
I understand that it is possible to just pass the parameter in the URL and guess by the total count that the last page has been reached. But I am referring to the `getAll()` method of Spinen\ConnectWise\Api\Client. If I use `getAll`, the pageSize changes from 25 to 100, which is why your example works. but lets say I want to call `$client->getAll('company/companies')` . In my database, this is thousands of records, and as the code is, only the first 100 are returned, and here is why:

* I call `$client->getAll('company/companies')
* Client::__call is executed with arguments `getAll` and `['company/companies']`
* Because the verb is `getAll`, the code on line 157 of the Client class erroneously  sets Client::page to `true` and `$verb` to 'get'
* The request is then passed on to the Client::request method and then eventually to the guzzle client.
* Before the guzzle client executes the request, the token needs to be populated
* This triggers a call to Client::__call with arguments `get` and `['system/members/legattd/tokens']`
* Because of the code on line 157, Client::page is now set to 0 because the method is no longer 'getAll'
* In the Client::request method, on line 455, there is an if statement to see if Client::page is truthy
* Because the token had to be fetched, Client::page is now equal to 0 and therefor no longer truthy, so the first result set is returned, rather than continuing on to the while loop on line 461 that would automatically iterate the pages using the Link header provided by CW.

In the fix I propose, Client::page is never set to 0, so if the Link header indicates that there are more pages, they will be iterated. This may result in odd behavior, but nothing that would break an app. I know my solution is just a bandaid, because as you can see, there are state issues with the Client class that would require more refactoring to resolve.

**Comment by jimmypuckett** [comment]
@dlegatt This is a very interesting catch.  We have never ran into this because everywhere that we are using the package, we have a first call to get a system value, which populates the token.  Then our `getAll` calls aways work.  It was not until I did it twice directly from `tinker` that I can see the issue...

```php
$ docker-compose exec web php artisan tinker
Psy Shell v0.9.9 (PHP 7.2.14 — cli) by Justin Hileman
>>> $cw = app(Spinen\ConnectWise\Api\Client::class)
=> Spinen\ConnectWise\Api\Client {#3558}
>>> count($cw->getAll('company/companies'))
=> 100 // first call limited to the 100
>>> count($cw->getAll('company/companies'))
=> 7194 // second call has the token, so full call
>>>
```

As you can see, the first call hit the bug you discovered with the increased `pageSize` of a 100 (which as you stated my member example worked), but then my second call got all of the records.  

We will look into this & see about fixing it & what we are missing on our tests.

Thanks for flushing this out.

**Comment by jimmypuckett** [comment]
Adding @ssfinney

**Comment by ssfinney** [comment]
@dlegatt Thanks for bringing this up! I've pushed the fix you suggested and tagged it at version 5.5.5.
